prow/git/git.go

```diff
@@ -32,26 +31,176 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type gitExecutor interface {
+	run(args ...string) ([]byte, error)
+	runWithRetries(retries int, args ...string) ([]byte, error)
+	remote() string
+}
+
+func newLiteralGitExecutor(dir, git string, remoteURL *url.URL) gitExecutor {
+	return &literalGitExecutor{
+		logger: logrus.WithFields(logrus.Fields{
+			"client": "git",
+			"dir":    dir,
+		}),
+		dir:       dir,
+		git:       git,
+		remoteURL: remoteURL,
+	}
+}
+
+type literalGitExecutor struct {
+	// logger will be used to log git operations
+	logger *logrus.Entry
+	// rootCacheDir is the location of the git cache for this repo.
+	dir string
+	// git is the path to the git binary.
+	git string
+	// base is the base path for git clone calls. For users it will be set to
+	// GitHub, but for tests set it to a directory with git repos.
+	remoteURL *url.URL
+}
+
+// run will pass the args to git and run the command
+func (e *literalGitExecutor) run(args ...string) ([]byte, error) {
+	var b []byte
+	var err error
+	logger := e.logger.WithField("args", strings.Join(args, " "))
+	c := exec.Command(e.git, args...)
+	c.Dir = e.dir
+	b, err = c.CombinedOutput()
+	if err != nil {
+		logger.WithError(err).Warningf("Running command returned error %v with output %s.", err, string(b))
+	} else {
+		logger.Info("Ran command.")
+	}
+
+	return b, err
+}
+
+// run will pass the args to git and run the command, retrying the specified
+// number of times with an exponential backoff in between
+func (e *literalGitExecutor) runWithRetries(retries int, args ...string) ([]byte, error) {
+	var b []byte
+	var err error
+	sleepyTime := time.Second
+	logger := e.logger.WithField("args", strings.Join(args, " "))
+	for i := 0; i < retries; i++ {
+		b, err = e.run(args...)
+		if err != nil {
+			logger.WithField("attempt", i).Warning("Running command failed, retrying...")
+			time.Sleep(sleepyTime)
+			sleepyTime *= 2
+			continue
+		}
+		break
+	}
+	if err != nil {
+		logger.Error("Failed to run command.")
+	}
+	return b, err
+}
+
+// remote exposes the remote URL for use in commands against the remote
+func (e *literalGitExecutor) remote() string {
+	return e.remoteURL.String()
+}
+
+type cacheInteractor interface {
```

Comment:
Could you please elaborate on the purpose of the cache abstraction layer? I seems like the existing implementation is quite hermetic already...

Reply:
I haven't added tests for this one yet but in general these functional interfaces let us inject dependencies into higher-level structures to test their exported methods

Reply:
I mean it feels like cache is hermetic so it is safe to use for testing purposes. IMHO, mocking it out in higher-level objects would actually reduce the test quality since now plugin tests are also serve as git package integration tests.

Reply:
Not sure what you mean. We'd mock it out in tests specifically targeting this behavior. I've made no changes to the `localgit` implementation